File: demos/patchy_disc.cpp
/*
  Copyright (c) 2015 Lester Hedges <lester.hedges+vmmc@gmail.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#ifdef ISOTROPIC
#error patchy_disc.cpp cannot be compiled against isotropic VMMC library!
#endif

#include "Demo.h"
#include "VMMC.h"

int main(int argc, char** argv)
{
    // Simulation parameters.
    unsigned int dimension = 2;                     // dimension of simulation box
    unsigned int nParticles = 1000;                 // number of particles
    double interactionEnergy = 8.0;                 // pair interaction energy scale (in units of kBT)
    double interactionRange = 0.1;                  // diameter of patch (in units of particle diameter)
    double density = 0.2;                           // particle density
    double baseLength;                              // base length of simulation box
    unsigned int maxInteractions = 3;               // maximum number of interactions per particle (number of patches)

    // Data structures.
    std::vector<Particle> particles(nParticles);    // particle container
    CellList cells;                                 // cell list
    bool isIsotropic[nParticles];                   // whether the potential of each particle is isotropic

    // Resize particle container.
    particles.resize(nParticles);

    // Work out base length of simulation box (particle diameter is one).
    if (dimension == 2) baseLength = std::pow((nParticles*M_PI)/(2.0*density), 1.0/2.0);
    else baseLength = std::pow((nParticles*M_PI)/(6.0*density), 1.0/3.0);

    std::vector<double> boxSize;
    for (unsigned int i=0;i<dimension;i++)
        boxSize.push_back(baseLength);

    // Initialise simulation box object.
    Box box(boxSize);

    // Initialise input/output class,
    InputOutput io;

    // Create VMD script.
    io.vmdScript(boxSize);

    // Initialise cell list.
    cells.setDimension(dimension);
    cells.initialise(box.boxSize, 1 + 0.5*interactionRange);

    // Initialise random number generator.
    MersenneTwister rng;

    // Initialise particle initialisation object.
    Initialise initialise;

    // Generate a random particle configuration.
    initialise.random(particles, cells, box, rng);

    // Initialise the patchy disc model.
    PatchyDisc patchyDisc(box, particles, cells,
        maxInteractions, interactionEnergy, interactionRange);

    // Initialise data structures needed by the VMMC class.
    double coordinates[dimension*nParticles];
    double orientations[dimension*nParticles];

    // Copy particle coordinates and orientations into C-style arrays.
    for (unsigned int i=0;i<nParticles;i++)
    {
        for (unsigned int j=0;j<dimension;j++)
        {
            coordinates[dimension*i + j] = particles[i].position[j];
            orientations[dimension*i + j] = particles[i].orientation[j];
        }

        // Set all particles as anisotropic.
        isIsotropic[i] = false;
    }

    // Initalise VMMC object.
    vmmc::VMMC vmmc(&patchyDisc, nParticles, dimension, coordinates, orientations,
        0.15, 0.2, 0.5, 0.5, maxInteractions, &boxSize[0], isIsotropic, false);

    // Execute the simulation.
    for (unsigned int i=0;i<1000;i++)
    {
        // Increment simulation by 1000 Monte Carlo Sweeps.
        vmmc += 1000*nParticles;

        // Append particle coordinates to an xyz trajectory.
        if (i == 0) io.appendXyzTrajectory(dimension, particles, true);
        else io.appendXyzTrajectory(dimension, particles, false);

        // Report.
        printf("sweeps = %9.4e, energy = %5.4f\n", ((double) (i+1)*1000), patchyDisc.getEnergy());
    }

    std::cout << "\nComplete!\n";

    // We're done!
    return (EXIT_SUCCESS);
}

File: demos/python/demo/patchydisc.py
# Copyright (c) 2015-2016 Lester Hedges <lester.hedges+vmmc@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""
@package demo
@author Lester Hedges
@brief Derived class for implementing a simple two-dimension patchy disc potential.
"""

import math
from model import Model
from operator import mul, sub

class PatchyDisc(Model):
    """ Derived class for implementing a simple two-dimension patchy disc potential. """

    def __init__(self,
                 box,
                 particles,
                 cells,
                 max_interactions,
                 interaction_energy,
                 interaction_range):

        # Call base class constructor.
        Model.__init__(self, box, particles, cells, max_interactions, interaction_energy, interaction_range)

        # Work out the angle between patches.
        self.patch_separation = 2*math.pi / max_interactions

        # Create rotation matrix lookup tables.
        self.cos_theta = [0] * max_interactions
        self.sin_theta = [0] * max_interactions

        for x in range(0, max_interactions):
            self.cos_theta[x] = math.cos(x*self.patch_separation)
            self.sin_theta[x] = math.sin(x*self.patch_separation)

    def compute_pair_energy(self, particle1, position1, orientation1, particle2, position2, orientation2):
        """ Compute pair interaction energy between two discs. """

        # Calculate particle separation.
        sep = map(sub, position1, position2)

        # Compute minimum image.
        self.box.minimum_image(sep)

        # Compute squared norm of separation vector.
        norm_sqd = sum(map(mul, sep, sep))

        # Discs overlap.
        if norm_sqd < 1:
            return float("inf")

        # Zero energy sum.
        energy = 0

        # Initialise disc coordinate arrays.
        coord1 = [0] * 2
        coord2 = [0] * 2

        # Test interactions between all patch pairs.
        for i in range(0, self.max_interactions):

            # Compute position of patch i on first disc
            coord1[0] = position1[0] + 0.5*(orientation1[0]*self.cos_theta[i] - orientation1[1]*self.sin_theta[i])
            coord1[1] = position1[1] + 0.5*(orientation1[0]*self.sin_theta[i] + orientation1[1]*self.cos_theta[i])

            # Enforce periodic boundaries.
            self.box.periodic_boundaries(coord1)

            for j in range(0, self.max_interactions):

                # Compute position of patch j on second disc
                coord2[0] = position2[0] + 0.5*(orientation2[0]*self.cos_theta[i] - orientation2[1]*self.sin_theta[i])
                coord2[1] = position2[1] + 0.5*(orientation2[0]*self.sin_theta[i] + orientation2[1]*self.cos_theta[i])

                # Enforce periodic boundaries.
                self.box.periodic_boundaries(coord1)

                # Calculate patch separation.
                sep = map(sub, coord1, coord2)

                # Compute minimum image.
                self.box.minimum_image(sep)

                # Compute squared norm of separation vector.
                norm_sqd = sum(map(mul, sep, sep))

                # Patches interact.
                if norm_sqd < self.squared_cutoff_distance:
                    energy -= self.interaction_energy

        return energy

    def compute_interactions(self, particle, position, orientation, interactions):
        """ Compute the number of interactions between two discs. """

        # Zero number of interactions.
        num_interactions = 0

        # Loop over all nearest neighbour cells.
        for cell in self.cells.neighbours[self.particles[particle].cell]:

            # Loop over all particles in cell.
            for x in range(0, self.cells.tally[cell]):

                neighbour = self.cells.particles[cell][x]

                # Make sure particles are different.
                if neighbour != particle:

                    # Calculate pair energy.
                    energy = self.compute_pair_energy(particle, position, orientation, neighbour,
                        self.particles[neighbour].position, self.particles[neighbour].orientation)

                    # Discs interact.
                    if energy < 0:

                        # Check maxmium overlaps.
                        if num_interactions == self.max_interactions:
                            raise ValueError('Maximum number of interactions exceeded.')
                            raise SystemExit

                        interactions[num_interactions] = neighbour
                        num_interactions += 1

        return num_interactions
